src/main/java/one/util/streamex/PrefixOps.java

```diff
@@ -1,302 +1,478 @@
+            super(source, (a, b) -> a == NONE ? b : op.apply(a, b));
+            this.localOp = op;
+        }
+        
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            if (!source.tryAdvance(this)) {
+                return false;
+            }
+            action.accept(acc);
+            return true;
+        }
+        
+        @Override
+        public void forEachRemaining(Consumer<? super T> action) {
+            if (accRef == null) {
+                source.forEachRemaining(next -> action.accept(acc = op.apply(acc, next)));
+            } else {
+                @SuppressWarnings("unchecked")
+                T[] buf = (T[]) new Object[BUF_SIZE];
+                source.forEachRemaining(next -> {
+                    if (idx == 0) {
+                        buf[idx++] = next;
+                    } else {
+                        T prev = buf[idx - 1];
+                        buf[idx++] = localOp.apply(prev, next);
+                        if (idx == buf.length) {
+                            drain(action, buf);
+                            idx = 0;
+                        }
+                    }
+                });
+                if (idx > 0)
+                    drain(action, buf);
+            }
+        }
+
+        private void drain(Consumer<? super T> action, T[] buf) {
+            T last = buf[idx - 1];
+            T acc = accRef.getAndAccumulate(last, op);
+            if (acc != NONE) {
+                for (int i = 0; i < idx; i++) {
+                    action.accept(localOp.apply(buf[i], acc));
+                }
+            } else {
+                for (int i = 0; i < idx; i++) {
+                    action.accept(buf[i]);
+                }
+            }
+        }
+
+        @Override
+        public void accept(T next) {
+            if (accRef == null) {
+                acc = op.apply(acc, next);
+            } else {
+                acc = accRef.accumulateAndGet(next, op);
+            }
+        }
+    }
+    
+    static final class OfUnordInt extends AbstractIntSpliterator implements IntConsumer {
+        private final IntBinaryOperator op;
+        private final Spliterator.OfInt source;
+        int idx = 0;
+        private boolean started;
+        AtomicInteger accRef;
+        private int acc;
+        
+        OfUnordInt(Spliterator.OfInt source, IntBinaryOperator op) {
+            super(source.estimateSize(), source.characteristics() & (ORDERED | IMMUTABLE | CONCURRENT | SIZED | NONNULL));
+            this.source = source;
+            this.op = op;
+        }
+        
+        @Override
+        public Spliterator.OfInt trySplit() {
+            if (started) {
+                return null;
+            }
+            OfInt prefix = source.trySplit();
+            if (prefix == null) {
+                return null;
+            }
+            if (accRef == null) {
+                accRef = new AtomicInteger();
```

Comment:
Unfortunately, this implementation is incorrect and cannot be accepted. Here's a simple test that fails:
`assertEquals(1024, IntStreamEx.constant(2, 10).prefix((a, b) -> a*b).max().getAsInt());` (committed it)
It's not correct to feed the user-supplied operation with zero because it's not guaranteed that a zero would act as an identity element for any operation. OfUnordRef solves this using a sentinel, but it's not possible to have a sentinel inside AtomicInteger. It's really complex to make this correctly and faster than the current implementation. Likely, it's possible to do this with a StampedLock, but I have little experience with this.

Reply:
This was a major fail! Thanks for admitting it.
Managed to fix it keeping performance boost:

Benchmark                                       (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelOrdered     100000  thrpt   25  2586,580 ±  98,957  ops/s
PrefixIntStreamBenchmark.parallelUnordered   100000  thrpt   25  4513,516 ± 408,134  ops/s
PrefixLongStreamBenchmark.parallelOrdered    100000  thrpt   25  2321,801 ±   5,534  ops/s
PrefixLongStreamBenchmark.parallelUnordered  100000  thrpt   25  3803,353 ± 628,167  ops/s

Reply:
Just realized I`ve introduced race condition with this fix.
Already repaired. Performance is still good:

Benchmark                                       (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelOrdered     100000  thrpt   25  2646,411 ±  91,600  ops/s
PrefixIntStreamBenchmark.parallelUnordered   100000  thrpt   25  4571,769 ± 286,287  ops/s
PrefixLongStreamBenchmark.parallelOrdered    100000  thrpt   25  2325,392 ±  14,375  ops/s
PrefixLongStreamBenchmark.parallelUnordered  100000  thrpt   25  3882,683 ± 204,037  ops/s

Reply:
What benchmark shows on your machine for a similar test for reference type? How many cores does your machine have? What are results for non-parallel execution?

Reply:
>How many cores does your machine have? 

8

>What are results for non-parallel execution?

PrefixIntStreamBenchmark.sequentialOrdered       100000  thrpt   25  1241,651 ± 17,490  ops/s
PrefixIntStreamBenchmark.sequentialUnordered     100000  thrpt   25  1585,448 ± 51,075  ops/s
PrefixLongStreamBenchmark.sequentialOrdered      100000  thrpt   25  1217,639 ± 20,058  ops/s
PrefixLongStreamBenchmark.sequentialUnordered    100000  thrpt   25  1390,811 ± 21,555  ops/s

Should I add these benchmarks to the repository too?

>What benchmark shows on your machine for a similar test for reference type?

PrefixObjectStreamBenchmark.parallelOrdered      100000  thrpt   25   319,191 ±  2,121  ops/s
PrefixObjectStreamBenchmark.parallelUnordered    100000  thrpt   25  1501,645 ± 37,583  ops/s
PrefixObjectStreamBenchmark.sequentialOrdered    100000  thrpt   25   814,598 ± 12,798  ops/s
PrefixObjectStreamBenchmark.sequentialUnordered  100000  thrpt   25   702,164 ±  7,094  ops/s

Should I add these benchmarks to the repository too?